### DISCUSSION TITLE
Plotting of results from pyswmm

### DISCUSSION THREAD

**Discussion by Dinho4** [pyswmm/pyswmm#332] [Q&A] [answered]
After running a simulation in pyswmm; how do you then the plot your results, specifically for volume time-series of storage units?

**Comment by bemcdonnell** [comment]
@Dinho4 check out matplotlib or plotly

**Comment by floriandorre** [comment]
You can either do it with some code by having a look at [swmmtoolbox](https://github.com/timcera/swmmtoolbox) and combine it with any plot package of python or you can do it using SWMM GUI but you first need to update the ini file linked to the output (https://github.com/OpenWaterAnalytics/pyswmm/issues/224#issuecomment-1042941862)

**Comment by jennwuu** [comment, marked as answer]
You can also check out this tutorial here: https://medium.com/@wuu.jennifer/pyswmm-tutorial-how-to-visualize-swmm-model-data-using-pandas-and-matplotlib-75aee5580dec

You can also check out the Output module in PySWMM to access output data. 
https://pyswmm.readthedocs.io/en/v1.1.1/tutorial/tutorial.html#access-swmm-output-binary-file

**Reply by Dinho4** [reply]
I get the following error when I try to visualize swmm model data using pandas and matplotlib:
import pandas as pd
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'pandas'

**Reply by jennwuu** [reply]
Hey @Dinho4 you need to install pandas to use pandas.

Here is a 10minute tutorial for using pandas.
https://pandas.pydata.org/pandas-docs/stable/user_guide/10min.html

**Reply by bemcdonnell** [reply]
@Dinho4 we are more than happy to help with larger Python concepts.  However, if in exchange, you could spend some time on StackOverflow or google the errors that you are having before reaching out, we can better keep this discussion be focused on PySWMM rather than Python. :)

**Reply by Dinho4** [reply]
Noted @bemcdonnell; when I try to access output data using Output module in Pyswmm, I get the following error;
>>> from pyswmm import Output
>>> with Output('C:/Users/Malesela/Downloads/CIV5000WProject.out') as out:
...     print(len(out.subcatchments))
...     print(len(out.nodes))
...     print(len(out.links))
...
swig/python detected a memory leak of type 'struct Handle *', no destructor found.
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Malesela\AppData\Roaming\Python\Python38\site-packages\pyswmm\output.py", line 203, in __enter__
    self.open()
  File "C:\Users\Malesela\AppData\Roaming\Python\Python38\site-packages\pyswmm\output.py", line 178, in open
    output.open(self.handle, self.binfile)
  File "C:\Users\Malesela\AppData\Roaming\Python\Python38\site-packages\swmm\toolkit\output.py", line 108, in open
    return _output.open(p_handle, path)
Exception: File Error 434: unable to open binary output file

**Reply by jennwuu** [reply]
@Dinho4 
Try adding r infront of your file path. 

e.g. 

```
with Output(r'C:/Users/Malesela/Downloads/CIV5000WProject.out') as out:
```

source: https://stackoverflow.com/questions/33729045/what-does-an-r-represent-before-a-string-in-python#:~:text=r%20means%20the%20string%20will,are%20left%20in%20the%20string.

**Reply by Dinho4** [reply]
I still get the same results even when I added r in front of my file path:
>>> from pyswmm import Output
>>> with Output(r'C:/Users/Malesela/Downloads/CIV5000WProject.out') as out:
...     print(len(out.subcatchments))
...     print(len(out.nodes))
...     print(len(out.links))
...
swig/python detected a memory leak of type 'struct Handle *', no destructor found.
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Malesela\AppData\Roaming\Python\Python38\site-packages\pyswmm\output.py", line 203, in __enter__
    self.open()
  File "C:\Users\Malesela\AppData\Roaming\Python\Python38\site-packages\pyswmm\output.py", line 178, in open
    output.open(self.handle, self.binfile)
  File "C:\Users\Malesela\AppData\Roaming\Python\Python38\site-packages\swmm\toolkit\output.py", line 108, in open
    return _output.open(p_handle, path)
Exception: File Error 434: unable to open binary output file

**Reply by bemcdonnell** [reply]
@Dinho4 are there more lines in your script beyond trying to open the *OUT file?  I'm wondering if the *OUT file might be open in a separate process.

**Reply by Dinho4** [reply]
No; after my simulations, the only thing I have been trying since is to access the *OUT file but if that's the case that might be open in a separate process, how then can I close it or what should be done?

**Reply by Dinho4** [reply]
I have two OUT*files namely CIV5000WProject.out(1,4GB in size) which is more recent and CIV5000W Simulation Project.out(388KB in size) which is older; I am able to access the output binary file for the 388KB OUT*file but not for the 1,4GB which gives me the errors as discussed above. Could this mean that my machine which is [Intel(R) Core(TM) i3-6006U CPU @ 2.00GHz   2.00 GHz] is compatible for only small OUT*files?
